File: 1306-jump-game-iii/1306-jump-game-iii.py
class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:
        
        def bfs(x):
            visited, queue, n = set(), deque(), len(arr)
            inbound = lambda x: 0 <= x < n
            queue.append(x)
            visited.add(x)
            while queue:
                graph = []
                cur = queue.popleft()
                if not arr[cur]:
                    return True
                if inbound(cur+arr[cur]):
                    graph.append(cur+arr[cur])
                if inbound(cur-arr[cur]):
                    graph.append(cur-arr[cur])
                    
                for neighbor in graph:
                    if neighbor not in visited:
                        queue.append(neighbor)
                        visited.add(neighbor)
                        
            return False
            
        return bfs(start)    

File: 1306-jump-game-iii/1306-jump-game-iii.cpp
class Solution {
public:
    bool canReach(vector<int>& arr, int start) {
        queue<int> q;
        q.push(start);
        set<int> vis;
        vis.insert(start);
        function<bool(int)> inbound = [&](int x) {
            return (x >= 0 && x < arr.size());
        };
        while (!q.empty()) {
            int i = q.front();
            q.pop();
            if (arr[i] == 0)
                return true;
            if (!vis.count(i + arr[i]) && inbound(i + arr[i])) {
                vis.insert(i + arr[i]);
                q.push(i + arr[i]);
            }
            if (!vis.count(i - arr[i]) && inbound(i - arr[i])) {
                vis.insert(i - arr[i]);
                q.push(i - arr[i]);
            }
        }
        return false;
    }
};